lib/async.js

```diff
@@ -261,31 +275,39 @@ module.exports = function resolve(x, options, callback) {
     }
 
     function processDirs(cb, dirs) {
-        if (dirs.length === 0) return cb(null, undefined);
-        var dir = dirs[0];
+        iterate(0);
 
-        isDirectory(dir, isdir);
+        function iterate(i) {
+            if (i === dirs.length) return cb(null, undefined);
+            var dir = dirs[i];
 
-        function isdir(err, isdir) {
-            if (err) return cb(err);
-            if (!isdir) return processDirs(cb, dirs.slice(1));
-            var file = path.join(dir, x);
-            loadAsFile(file, opts.package, onfile);
-        }
+            isDirectory(path.dirname(dir), function (err, status) {
```

Comment:
why the extra path.dirname call? shouldn't the `dirs` list already have that done?

Reply:
`dirs` may point to files, for example if you do resolve something like `foo/package.json` then `dirs` will be `['/n_m/foo/package.json']` (that makes `dir` a misnomer, do you want me to rename it into `entry` or similar?).

Reply:
i mean, yes, but in that case `foo` should be the entry that's a directory and `package.json` should be the entry that's not.

Reply:
That's why I call `dirname`: since `dir` is `/n_m/foo/package.json`, the isDir check has to be done against its parent. Even if it was `/n_m/foo` (which would happen if the request was just `foo`), the right behavior would be to check against `/n_m` (since `foo` is allowed to be a file per the Node resolution rules), so the `dirname` would still stand.